File: QueueWithArray/queue.py
class Queue:
  def __init__(self, capacity):
    self.front = self.size = 0
    self.rear = capacity -1
    self.Q = [None]*capacity
    self.capacity = capacity

  def EnQueue(self, item):
    if self.isFull():
      return
    self.rear = (self.rear + 1) % (self.capacity)
    self.Q[self.rear] = item
    self.size = self.size + 1

  def DeQueue(self):
    if self.isEmpty():
      return
    self.front = (self.front + 1) % (self.capacity)
    self.size = self.size -1
    
  def que_front(self):
    if self.isEmpty():
      print("Queue is empty")
    print("Front item is", self.Q[self.front])
    
  def que_rear(self):
    if self.isEmpty():
      print("Queue is empty")
    print("Rear item is", self.Q[self.rear])

  def isFull(self):
    return self.size == self.capacity

  def isEmpty(self):
    return self.size == 0


if __name__ == '__main__':
  queue = Queue(5)
  queue.EnQueue(1)
  queue.EnQueue(2)
  queue.EnQueue(3)
  queue.EnQueue(4)
  queue.DeQueue()
  queue.que_front()
  queue.que_rear()

File: QueueWithArray/queue.cpp
class Queue
{
public:
  int front, rear, size;
  unsigned capacity;
  int* array;
};

Queue* createQueue(unsigned capacity)
{
  Queue* queue = new Queue();
  queue->capacity = capacity;
  queue->front = queue->size = 0;
  queue->rear = capacity - 1;
  queue->array = new int[(queue->capacity * sizeof(int))];
  return queue; 
}

void enqueue(Queue* queue, int item)
{
  if (isFull(queue))
    return;
  queue->rear = (queue->rear + 1) % queue->capacity;
  queue->array[queue->rear] = item;
  queue->size = queue->size + 1;
}

int dequeue(Queue* queue)
{
  if (isEmpty(queue))
    return INT_MIN;
  int item = queue->array[queue->front];
  queue->front = (queue->front + 1) % queue->capacity;
  queue->size = queue->size - 1;
  return item;
}

int front(Queue* queue)
{
  if (isEmpty(queue))
    return INT_MIN;
  return queue->array[queue->front];
}

int rear(Queue* queue)
{
  if (isEmpty(queue))
    return INT_MIN;
  return queue->array[queue->rear];
}

int isFull(Queue* queue) { return (queue->size == queue->capacity); }

int isEmpty(Queue* queue) { return (queue->size == 0); }


int main()
{
  Queue* queue = createQueue(10);
  enqueue(queue, 1);
  enqueue(queue, 2);
  enqueue(queue, 3);
  enqueue(queue, 4);
  dequeue(queue);
  front(queue);
  rear(queue);
}
